keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2168 @@
+
+## Summary
+
+This KEP proposes adding a declarative API to manage node maintenance. This API can be used to
+implement additional capabilities around node draining.
+
+## Motivation
+
+The goal of this KEP is to analyze and improve node maintenance in Kubernetes.
+
+Node maintenance is a request from a cluster administrator to remove all pods from a node(s) so
+that it can be disconnected from the cluster to perform a software upgrade (OS, kubelet, etc.),
+hardware or firmware upgrade, or simply to remove the node as it is no longer needed.
+
+Kubernetes has existing support for this use case in the following way with `kubectl drain`:
+1. There are running pods on node A, some of which are protected with PodDisruptionBudgets (PDB).
+2. Set the node `Unschedulable` (cordon) to prevent new pods from being scheduled there.
+3. Evict (default behavior) pods from node A by using the eviction API (see [kubectl drain worklflow](https://raw.githubusercontent.com/kubernetes/website/f2ef324ac22e5d9378f2824af463777182817ca6/static/images/docs/kubectl_drain.svg)).
+4. Proceed with the maintenance and shut down or restart the node.
+5. On platforms and nodes that support it, the kubelet will try to detect the imminent shutdown and
+   then attempt to perform a [Graceful Node Shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown):
+   - delay the shutdown pending graceful termination of remaining pods
+   - terminate remaining pods in reverse priority order (see [pod-priority-graceful-node-shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#pod-priority-graceful-node-shutdown))
+
+The main problem is that the current approach tries to solve this in an application agnostic way
+and will simply attempt to remove all the pods currently running on the node. Since this approach
+cannot be applied generically to all pods, the Kubernetes project has defined special
+[drain filters](https://github.com/kubernetes/kubernetes/blob/56cc5e77a10ba156694309d9b6159d4cd42598e1/staging/src/k8s.io/kubectl/pkg/drain/filters.go#L153-L162)
+that either skip groups of pods or an admin has to consent to override those groups to be either
+skipped or deleted. This means that without knowledge of all the underlying applications on the
+cluster, the admin has to make a potentially harmful decision.
+
+From an application owner or developer perspective, the only standard tool they have is
+a PodDisruptionBudget. This is sufficient in a basic scenario with a simple multi-replica
+application. The edge case applications, where this does not work are very important to
+the cluster admin, as they can block the node drain. And, in turn, very important to the
+application owner, as the admin can then override the pod disruption budget and disrupt their
+sensitive application anyway.
+
+List of cases where the current solution is not optimal:
+
+1. Without extra manual effort, an application running with a single replica has to settle for
```

Comment:
Some applications have a general ability (e.g. deployments) to scale up first. So they do not have to experience the downtime if configured correctly. This can be generalized to any number of replicas. There are more details in the linked issues.